src/Commands/Plugin/PluginBaseCommand.php

```diff
@@ -0,0 +1,241 @@
+abstract class PluginBaseCommand extends TerminusCommand
+{
+    /**
+     * PluginBaseCommand constructor.
+     */
+    public function __construct()
+    {
+    }
+
+    /**
+     * Get plugin projects.
+     *
+     * @param string $plugins_dir Plugins directory
+     * @return array Plugin projects
+     */
+    protected function getPluginProjects($plugins_dir)
+    {
+        $projects = array();
+        $finder = new Finder();
+        $finder->files()->in($plugins_dir);
+        foreach ($finder as $file) {
+            $path = $file->getRelativePath();
+            // Get the parent path only.
+            if (!strpos($path, DIRECTORY_SEPARATOR)) {
+                // Make sure the path is unique.
+                if (!in_array($path, $projects)) {
+                    $projects[] = $path;
+                }
+            }
+        }
+        return $projects;
+    }
+
+    /**
+     * Get the plugin directory.
+     *
+     * @param string $plugin Plugin name
+     * @return string Plugin directory
+     */
+    protected function getPluginDir($plugin = '')
+    {
+        $plugins_dir = $this->getConfig()->get('plugins_dir');
+        // Create the directory if it doesn't already exist.
+        if (!is_dir("$plugins_dir")) {
+            mkdir("$plugins_dir", 0755, true);
+        }
+        return $plugins_dir . DIRECTORY_SEPARATOR . $plugin;
+    }
+
+    /**
+     * Get the plugin installation method.
+     *
+     * @param string Plugin name
+     * @return string Plugin installation method
+     */
+    protected function getInstallMethod($plugin)
+    {
+        $plugin_dir = $this->getPluginDir($plugin);
+        $git_dir = $plugin_dir . DIRECTORY_SEPARATOR . '.git';
+        if (is_dir("$git_dir")) {
+            return 'git';
+        }
+        $composer_json = $plugin_dir . DIRECTORY_SEPARATOR . 'composer.json';
+        return file_exists($composer_json) ? 'composer' : 'unknown';
+    }
+
+    /**
+     * Get the currently installed plugin version.
+     *
+     * @param string $plugin Path to plugin
+     * @return string Installed plugin version
+     */
+    protected function getInstalledVersion($plugin)
+    {
+        exec("cd \"$plugin\" && git describe --tags", $tags);
+        if (!empty($tags)) {
+            $version = array_pop($tags);
+        } else {
+            exec("cd \"$plugin\" && git rev-parse --abbrev-ref HEAD", $branch);
+            $version = !empty($branch) ? array_pop($branch) : 'unknown';
+        }
+        return $version;
+    }
+
+    /**
+     * Get the latest available plugin version.
+     *
+     * @param string $plugin Path to plugin
+     * @return string Latest plugin version
+     */
+    protected function getLatestVersion($plugin)
+    {
+        // Get the Terminus major version.
+        $terminus_major_version = $this->getTerminusMajorVersion();
+        exec("cd \"$plugin\" && git fetch --all && git tag -l | grep ^[v$terminus_major_version] | sort -Vr | head -1", $tag);
```

Comment:
Writing tests now and caught a problem here. OS X's default `sort` command does not support `-V`.  Work-around is to apply the sorting on the `git tag` command. As of Git 2.0 (released in 2014), should be able to `git tag -l --sort=version:refname`

Reply:
I've allowed this to go for so long such that there is another contender for these commands, submitted the same day as this update. I'll need to take another look at the two side by side.
https://github.com/pantheon-systems/terminus/pull/1759

Reply:
Hello @TeslaDethray!  Feel free to use whatever method seems best.  I expected that we would need to change some things around before it was accepted.  If it helps any, I can give you access to edit this branch.  I gave @greg-1-anderson permission to make changes awhile ago.  Welcome back! :)

Reply:
It looks like #1759 is a continuation of this PR.

Reply:
Yep.  Would recommend consolidating, either by cherry-picking or merging.